### PR TITLE
Initial merge for tensorflow-serving-cluster

### PR DESCRIPTION
**Pull request by snormore**
A cluster of [tensorflow-serving](https://www.tensorflow.org/serving/) model servers on Kubernetes.

 - Deploy and run multiple `tensorflow-serving` instances on a Kubernetes cluster, with their Docker images fetching models and dictionaries at deploy time from a storage backend or URL
 - Classify images via HTTP(s) API proxy against any of the deployed models

[Build Status](https://travis-ci.org/triagemd/tensorflow-serving-cluster)

See https://github.com/triagemd/tensorflow-serving-cluster/blob/initial/README.md

----

# Getting started

... on an example project with model servers for `imagenet_mobilenet_v1`, `imagenet_inception_v3`, and `imagenet_resnet50`.

Install the [Docker Toolbox](https://www.docker.com/products/docker-toolbox) and [MiniKube](https://kubernetes.io/docs/tasks/tools/install-minikube/).

## Start up the cluster

```bash
git clone git@github.com:triagemd/tensorflow-serving-cluster.git
cd tensowflow-serving-cluster/example
source script/env
minikube start
tf-serving-cluster up
minikube dashboard
```

screen shot 2017-08-12 at 6 22 58 pm

## Query the cluster

```bash
time curl $(minikube service proxy-service --url)/imagenet_mobilenet_v1/classify?image=https://image.ibb.co/nu62ba/cat.jpg
time curl $(minikube service proxy-service --url)/imagenet_inception_v3/classify?image=https://image.ibb.co/nu62ba/cat.jpg
time curl $(minikube service proxy-service --url)/imagenet_resnet50/classify?image=https://image.ibb.co/nu62ba/cat.jpg
```

## Tear down the cluster:

```bash
tf-serving-cluster down
minikube down
```

----

# TODO pre-merge
 - [x] Add basic auth via environment
 - [x] Integrate a dictionary file via cluster config URL
 - [ ] Deploy to GKE with GC LB as the ingress controller
# TODO post-merge
 - [ ] Add SSL support via [kube-lego](https://github.com/jetstack/kube-lego)
 - [ ] Classify multiple images using the `images` param
 - [ ] CPU-optimized `tensorflow` in the `triage/tensorflow-serving-docker` Dockerfile
 - [ ] Performance analysis and benchmarks
 - [ ] Support dynamic model loading from a storage backend such as Google Storage
 - [ ] Better Docker and Python caching in CI
 - [x] Update k8s deployment on cluster `up` if it's already up
 - [ ] Selective cluster `down <component>`

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -1 +1,101 @@
...
+
+See https://github.com/triagemd/tensorflow-serving-cluster/blob/initial/example/README.md.
+
+
+## Start up the cluster
+
+```bash
+git clone git@github.com:triagemd/tensorflow-serving-cluster.git
+cd tensowflow-serving-cluster/example
+source script/env
```
Comment (stephensolis) [inline README.md]: `script/env` doesn't install dependencies, maybe you want to add `pip install -r requirements.txt` to the end?

File: README.md [hunk 2]
```diff
@@ -1 +1,101 @@
...
+----
+
+
+# Getting started
+
+... on a development and testing environment.
+
+ - Create your `cluster.yml`
+ - `pip install tensorflow-serving-cluster`
+ - `kubectl create secret generic gcloud --from-literal=GCLOUD_ACCOUNT=$(base64 .cache/gcloud-auth.json)`
```
Comment (stephensolis) [inline README.md]: Where's `gcloud-auth.json` from?

File: README.md [hunk 3]
```diff
@@ -1 +1,101 @@
...
+
+## Query the cluster
+
+```bash
+time curl $(minikube service proxy-service --url)/imagenet_mobilenet_v1/classify?image=https://image.ibb.co/nu62ba/cat.jpg
+time curl $(minikube service proxy-service --url)/imagenet_inception_v3/classify?image=https://image.ibb.co/nu62ba/cat.jpg
+time curl $(minikube service proxy-service --url)/imagenet_resnet50/classify?image=https://image.ibb.co/nu62ba/cat.jpg
+```
+
+![screen shot 2017-08-12 at 6 22 58 pm](https://user-images.githubusercontent.com/182290/29244650-5c480526-7f8b-11e7-9c87-8c5b50806aab.png)
```
Comment (stephensolis) [inline README.md]: Can you add instructions on how to open this interface?

File: docker/proxy/Dockerfile [hunk 4]
```diff
@@ -0,0 +1,21 @@
...
+RUN pip install --upgrade pip && \
+    pip install flask \
+                pyyaml \
+                google-cloud-storage \
+                kubernetes \
+                tensorflow_serving_client \
+                ml-tools
+
+ENV TENSORFLOW_SERVING_CLUSTER_VERSION 0.0.9
+RUN pip install tensorflow-serving-cluster==$TENSORFLOW_SERVING_CLUSTER_VERSION
```
Comment (stephensolis) [inline docker/proxy/Dockerfile]: How about just `pip install tensorflow-serving-cluster`?

File: example/README.md [hunk 5]
```diff
@@ -0,0 +1,32 @@
+Example [tensorflow-serving-cluster](https://github.com/triagemd/tensorflow-serving-cluster) project with model servers for `imagenet_mobilenet_v1`, `imagenet_inception_v3`, and `imagenet_resnet50`.
```
Comment (stephensolis) [inline example/README.md]: What's the separate `example/` folder for? Seems to be the same as the files in the root

File: example/script/env [hunk 6]
```diff
@@ -0,0 +1,10 @@
+#!/usr/bin/env bash
+VENV_DIR=".venv"
+if [ -z $NO_VENV ]; then
+  if [ ! -d $VENV_DIR ]; then
+    virtualenv .venv -p python3
+  fi
+  . $VENV_DIR/bin/activate
+fi
+pip install --upgrade pip
```
Comment (stephensolis) [inline example/script/env]: (this -> `script/env`)

File: tensorflow_serving_cluster/cluster.py [hunk 7]
```diff
@@ -0,0 +1,96 @@
...
+    def _read_config(self, path):
+        with open(path, 'r') as f:
+            return yaml.load(f)
+
+    def _build_proxy_service(self):
+        return KubernetesService({
+            'namespace': 'default',
+            'deployment_name': 'proxy-deployment',
+            'service_name': 'proxy-service',
+            'app_name': 'proxy-server',
```
Comment (stephensolis) [inline tensorflow_serving_cluster/cluster.py]: Can we change these to `tensorflow-serving-proxy` or something?
I was initially confused `proxy-server` was a Kubernetes thing.

File: tensorflow_serving_cluster/proxy_app.py [hunk 8]
```diff
@@ -0,0 +1,81 @@
...
+    @app.route('/<model_name>')
+    def model(model_name):
+        model = get_model(model_name)
+        return jsonify(model)
+
+    @app.route('/<model_name>/classify', methods=['GET', 'POST'])
+    def model_classify(model_name):
+        # Grab model.json from GS for the model_spec_name,
+        # and any overriding target_size
+        dictionary = None
```
Comment (stephensolis) [inline tensorflow_serving_cluster/proxy_app.py]: I think we might want to force dictionaries to be in the model spec, otherwise predictions get really hard to handle downstream

File: tests/test_proxy_app.py [hunk 9]
```diff
@@ -0,0 +1,68 @@
...
+def test_model_classify_no_image_file(client):
+    response = client.post('/imagenet_mobilenet_v1/classify')
+    assert response.status_code == 200
+    assert response.json == {'error': 'missing image file'}
+
+
+def test_model_classify_with_image_url(client):
+    response = client.get('/imagenet_mobilenet_v1/classify', query_string={'image': 'https://image.ibb.co/nu62ba/cat.jpg'})
+    assert response.status_code == 200
+    assert len(response.json) == 1000
```
Comment (stephensolis) [inline tests/test_proxy_app.py]: Seems a bit brittle, no?

File: example/README.md (hunk 5 above)
Comment (snormore) [inline example/README.md]: That's an example project that shows how a cluster can be created on it's own with `tensorflow-serving-cluster`. The files on the outside in the root of the project are used for development + testing of `tensorflow-serving-cluster` itself.

File: README.md (hunk 2 above)
Comment (snormore) [inline README.md]: That's the service account keyfile generated by GC. I'm adding encrypted secrets file/config support in `Cluster` today that will include that, it's needed for basic auth too. Also so that `Cluster` can decide not to have the k8s deployment look for that secret if it's not wanted, otherwise it errors (you might hit this if you're missing the GCLOUD_ACCOUNT secret for now).

File: tensorflow_serving_cluster/proxy_app.py (hunk 8 above)
Comment (snormore) [inline tensorflow_serving_cluster/proxy_app.py]: Yea 💯

File: tests/test_proxy_app.py (hunk 9 above)
Comment (snormore) [inline tests/test_proxy_app.py]: Because of the check for exactly 1000 classes? This test is here to check the proxy endpoint, it always serves the imagenet model in the test. I'm waiting to add dictionary expansion (pre-merge of this PR) to assert the contents of the response in more detail.

File: docker/proxy/Dockerfile (hunk 4 above)
Comment (snormore) [inline docker/proxy/Dockerfile]: Definitely preferable, but I don't want the Docker images to be unpredictable in that you're never sure which version it was deployed with. So right now a `distribute` will ship to pypi and dockerhub with the same versions.

File: README.md (hunk 1 above)
Comment (snormore) [inline README.md]: `example/script/env` installs dependencies, that's the one that's referenced here.

File: README.md (hunk 2 above)
Comment (stephensolis) [inline README.md]: You mean this one: https://cloud.google.com/storage/docs/authentication#generating-a-private-key?
Can you add a note on how to get one to https://github.com/triagemd/snap-serving/pull/1#discussion_r132846076 for our buckets?

File: tests/test_proxy_app.py (hunk 9 above)
Comment (stephensolis) [inline tests/test_proxy_app.py]: (I mean the JSON length check)

Comment (snormore) [inline tests/test_proxy_app.py]: Ah yea that's number of classes/predictions here, which is always 1k for imagenet models.

File: docker/proxy/Dockerfile (hunk 4 above)
Comment (stephensolis) [inline docker/proxy/Dockerfile]: The problem is that this and `setup.py` can get out of sync, and then `distribute` will actually ship an old version to dockerhub.

How about modifying `script/distribute-docker` to pull its version from `setup.py` instead?

Comment (snormore) [inline docker/proxy/Dockerfile]: Yea I like that, good idea

File: README.md (hunk 1 above)
Comment (stephensolis) [inline README.md]: Ah right oops!

File: tests/test_proxy_app.py (hunk 9 above)
Comment (stephensolis) [inline tests/test_proxy_app.py]: Oh OK thanks! Thought that was the JSON before decoding initially.

File: tests/test_proxy_app.py [hunk 10]
```diff
@@ -0,0 +1,131 @@
...
+
+
+def test_model_classify_with_image_url(client):
+    response = client.get('/imagenet_mobilenet_v1/classify',
+                          headers=auth_headers(),
+                          query_string={'image': 'https://image.ibb.co/nu62ba/cat.jpg'})
+    assert response.status_code == 200
+    assert len(response.json) == 1000
+    top_5 = sorted(response.json, reverse=True, key=lambda r: r[1])[:5]
+    assert top_5 == [
```
Comment (snormore) [inline tests/test_proxy_app.py]: 🐱 💯

File: tests/test_proxy_app.py [hunk 11]
```diff
@@ -0,0 +1,131 @@
...
+        ['tiger cat', 0.334694504737854],
+        ['Egyptian cat', 0.2851393222808838],
+        ['tabby, tabby cat', 0.15471667051315308],
+        ['kit fox, Vulpes macrotis', 0.03160465136170387],
+        ['lynx, catamount', 0.030886519700288773]
```
Comment (snormore) [inline tests/test_proxy_app.py]: 🐱💯

File: docker/proxy/entrypoint.sh [hunk 12]
```diff
@@ -0,0 +1,28 @@
+#!/usr/bin/env bash
+set -e
+
+# Auth with Google Cloud
+if [ ! -z $GCLOUD_ACCOUNT ]; then
+  mkdir -p /opt
+  echo ${GCLOUD_ACCOUNT} > /opt/gcloud-auth.json.base64
+  base64 -d /opt/gcloud-auth.json.base64 > /opt/gcloud-auth.json
+  gcloud auth activate-service-account --key-file=/opt/gcloud-auth.json ${GCLOUD_ACCOUNT_EMAIL}
```
Comment (stephensolis) [inline docker/proxy/entrypoint.sh]: I note this process isn't done on `script/test`, is that intentional?

File: script/distribute-docker [hunk 13]
```diff
@@ -1,11 +1,11 @@
 #!/usr/bin/env bash
 set -xe
 
-VERSION=$(grep "ENV TENSORFLOW_SERVING_CLUSTER_VERSION " docker/proxy/Dockerfile | cut -d' ' -f3)
+VERSION=$(grep "version=" setup.py | tr -d "version='" | tr -d ",'" | tr -d "[:space:]")
```
Comment (stephensolis) [inline script/distribute-docker]: 👍

File: docker/proxy/entrypoint.sh (hunk 12 above)
Comment (snormore) [inline docker/proxy/entrypoint.sh]: Yea this is only necessary for using the `gcloud` and `gsutil` tooling. The `gcloud` toolchain dependencies aren't really necessary in the proxy Dockerfile anymore, I'll clean that up.

File: tensorflow_serving_cluster/proxy_app.py [hunk 14]
```diff
@@ -0,0 +1,122 @@
...
+                if 'image' not in request.args:
+                    return jsonify({'error': 'missing image url'})
+                file.write(urlopen(request.args['image']).read())
+            image_data = load_image(file.name, model_spec.target_size, model_spec.preprocess_input)
+        if image_data is None:
+            return jsonify({'error': 'invalid image'})
+
+        response = client.make_prediction(image_data, 'image')
+        predictions = response['class_probabilities'][0].tolist()
+        predictions = list(zip(dictionary, predictions))
```
Comment (stephensolis) [inline tensorflow_serving_cluster/proxy_app.py]: I kind of liked this as `dict(zip(dictionary, predictions))`, why the change?

Comment (stephensolis) [inline tensorflow_serving_cluster/proxy_app.py]: Just for context, this was to keep the same class order as the model.
